src/main/java/org/folio/circulation/domain/LoanRepository.java

```diff
@@ -175,6 +176,39 @@ else if(loans.getTotalRecords() == 1) {
         .thenApply(result -> result.next(this::mapResponseToLoans));
   }
 
+  private CompletableFuture<Result<MultipleRecords<Loan>>> queryLoanStorageWithPagination(
+    Result<CqlQuery> statusQuery) {
+
+    CompletableFuture<Result<MultipleRecords<Loan>>> future = new CompletableFuture<>();
+    fetchNextPage(future, statusQuery, 0);
+    return future;
+  }
+
+  private void fetchNextPage(
+    CompletableFuture<Result<MultipleRecords<Loan>>> future,
+    Result<CqlQuery> statusQuery, int currentPage) {
+
+    getLoansByQuery(statusQuery, currentPage)
+      .thenApply(result -> {
+          int totalRecords = result.value().getTotalRecords();
+          int pageOffset = currentPage + 1;
+          if (totalRecords >  pageOffset * PAGE_LIMIT) {
+            fetchNextPage(future, statusQuery, pageOffset);
+          } else {
+            future.complete(result);
+          }
+          return result;
+        }
+      );
+  }
```

Comment:
Why has this been implemented here, when the multiple records fetcher can already fetch a set of records in batches? Was it because that functionality was not known about or inappropriate for this use?

Reply:
Has been the pagination already implemented? I haven't found implemented logic for that

Reply:
We use paging to prevent out-of-memory errors in both the server and the client.
MultipleRecordFetcher.java uses paging on the server side (mod-circulation-storage) but collects all results in memory on the client side (mod-circulation) and therefore may cause out-of-memory error.

Reply:
I don't see how LoanRepository.java avoids out-of-memory. It seem to (try to) collect all records of all pages and to return all of them.
Can we extend MultipleRecordFetcher.java to provide a method that runs a CqlQuery query and calls a page handler that returns the records of a single page only, and when that handler returns the records are removed from memory and it continues fetching the next page that calls that handler again? That way out-of-memory can be avoided.

Reply:
I would suggest moving the pagination to ItemLimitValidator. I will get 1000 loans per one iteration, filter them in the validator and save the number of loans. It will help to prevent out-of-memory. 
@julianladisch correct me if this approach is wrong please.

Reply:
As it was discussed with Khalilah Gambrell, we can limit loans by 10000. 
There is max count in description of items in the previous story: https://issues.folio.org/browse/UICIRC-390
For this count of loans no pagination is needed.